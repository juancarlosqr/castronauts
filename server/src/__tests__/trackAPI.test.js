const { ApolloServer, gql } = require('apollo-server');
const { createTestClient } = require('apollo-server-testing');
const casual = require('casual');
const typeDefs = require('../schema');
const resolvers = require('../resolvers');
const TrackAPI = require('../datasources/track-api');

casual.seed(4242);

const authorId = casual.uuid;

const mockTracksResponse = [
  {
    id: casual.uuid,
    title: casual.sentence,
    authorId: authorId,
    thumbnail: casual.url,
    length: casual.integer(1, 12),
    modulesCount: casual.integer(1, 5),
  },
  {
    id: casual.uuid,
    title: casual.sentence,
    authorId: authorId,
    thumbnail: casual.url,
    length: casual.integer(1, 12),
    modulesCount: casual.integer(1, 5),
  },
];

const mockAuthorResponse = {
  id: authorId,
  name: casual.name,
  photo: casual.url,
};

test('fetches tracks', async () => {
  const trackAPI = new TrackAPI();

  // create a test server to test against,
  // using our production typeDefs, resolvers, and dataSources
  const server = new ApolloServer({
    typeDefs,
    resolvers,
    dataSources: () => ({ trackAPI }),
  });

  // mock the dataSource's underlying fetch methods
  trackAPI.getTracks = jest.fn(() => mockTracksResponse);

  // create query
  const GET_LAUNCHES = gql`
    query getTracks {
      tracksForHome {
        id
        title
        length
        thumbnail
        modulesCount
      }
    }
  `;

  // use the test server to create a query function
  const { query } = createTestClient(server);

  // run query against the server and snapshot the output
  const res = await query({ query: GET_LAUNCHES });
  expect(res).toMatchSnapshot();
});

test('fetches tracks with author', async () => {
  const trackAPI = new TrackAPI();

  // create a test server to test against,
  // using our production typeDefs, resolvers, and dataSources
  const server = new ApolloServer({
    typeDefs,
    resolvers,
    dataSources: () => ({ trackAPI }),
  });

  // mock the dataSource's underlying fetch methods
  trackAPI.getTracks = jest.fn(() => mockTracksResponse);
  trackAPI.getAuthor = jest.fn(() => mockAuthorResponse);

  // create query
  const GET_LAUNCHES_AUTHOR = gql`
    query getTracksWithAuthor {
      tracksForHome {
        id
        title
        length
        thumbnail
        modulesCount
        author {
          id
          name
          photo
        }
      }
    }
  `;

  // use the test server to create a query function
  const { query } = createTestClient(server);

  // run query against the server and snapshot the output
  const res = await query({ query: GET_LAUNCHES_AUTHOR });
  expect(res).toMatchSnapshot();
});
