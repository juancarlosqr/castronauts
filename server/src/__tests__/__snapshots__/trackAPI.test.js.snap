// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches tracks 1`] = `
Object {
  "data": Object {
    "tracksForHome": Array [
      Object {
        "id": "4a1a1ce9-0544-4a7e-80d3-b509dc39e69e",
        "length": 7,
        "modulesCount": 2,
        "thumbnail": "http://www.Sadye.ca/",
        "title": "Hic aut ad maxime.",
      },
      Object {
        "id": "fabd5aae-877d-440d-bac6-f18cf3e4f994",
        "length": 7,
        "modulesCount": 4,
        "thumbnail": "http://www.Cheyenne.biz/",
        "title": "Recusandae sequi enim.",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`fetches tracks with author 1`] = `
Object {
  "data": Object {
    "tracksForHome": Array [
      Object {
        "author": Object {
          "id": "4cf3e745-d2ad-4885-ab28-f90277a2703f",
          "name": "Miss Adan Legros",
          "photo": "http://Dietrich.io/",
        },
        "id": "4a1a1ce9-0544-4a7e-80d3-b509dc39e69e",
        "length": 7,
        "modulesCount": 2,
        "thumbnail": "http://www.Sadye.ca/",
        "title": "Hic aut ad maxime.",
      },
      Object {
        "author": Object {
          "id": "4cf3e745-d2ad-4885-ab28-f90277a2703f",
          "name": "Miss Adan Legros",
          "photo": "http://Dietrich.io/",
        },
        "id": "fabd5aae-877d-440d-bac6-f18cf3e4f994",
        "length": 7,
        "modulesCount": 4,
        "thumbnail": "http://www.Cheyenne.biz/",
        "title": "Recusandae sequi enim.",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
